torch/csrc/jit/tensorexpr/kernel.cpp

```diff
@@ -1591,6 +1602,107 @@ void TensorExprKernel::bindInput(const torch::jit::Value* input) {
   }
 }
 
+namespace {
+
+// Remove all indices from axes positions.
+std::vector<VarHandle> squeezeIndices(
+    const ParameterList& indices,
+    const std::vector<size_t>& axes) {
+  std::vector<VarHandle> indices_squeezed;
+  for (size_t dim = 0; dim < indices.size(); ++dim) {
+    if (!std::count(axes.begin(), axes.end(), dim)) {
+      indices_squeezed.push_back(indices[dim]);
+    }
+  }
+  return indices_squeezed;
+}
+
+} // namespace
+
+Tensor* TensorExprKernel::computeSum(const torch::jit::Value* v) {
+  auto reduction_info = getReductionInfo(v->node());
+  return Reduce(
+      "sum",
+      reduction_info.outputDims,
+      Sum(),
+      [&](ParameterList& indices) {
+        const auto& axes = reduction_info.axes;
+        // "Squeeze" out indices inserted when keepdim is set.
+        auto indices_squeezed =
+            reduction_info.keepdim ? squeezeIndices(indices, axes) : indices;
+        TORCH_INTERNAL_ASSERT(axes.size() <= indices_squeezed.size());
+        // Move innermost indices into axes positions:
+        //   1. Fill the outermost indices first.
+        //   2. Insert the innermost indices into the correct axis position,
+        //   displacing the outermost indices as needed.
+        std::vector<ExprHandle> indices_exprs;
+        size_t i = 0;
+        for (; i < indices_squeezed.size() - axes.size(); ++i) {
+          indices_exprs.push_back(indices_squeezed[i]);
+        }
+        for (auto axis : axes) {
+          indices_exprs.insert(
+              indices_exprs.begin() + axis, indices_squeezed[i]);
+          ++i;
+        }
+        auto indexed = tensorOrConstant(v->node()->input(0), indices_exprs);
+        if (reduction_info.dtype) {
+          return Cast::make(*reduction_info.dtype, indexed);
+        } else {
+          return indexed;
+        }
+      },
+      reduction_info.reductionDims);
+}
+
+TensorExprKernel::ReductionInfo TensorExprKernel::getReductionInfo(
+    const torch::jit::Node* node) {
+  std::vector<size_t> axes;
+  bool keepdim = false;
+  auto sizes = sizesForValue(node->namedInput(attr::self));
+  const auto inputs = node->inputs();
+  if (inputs.size() > 2) {
+    // Canonicalize axes: wrap around, sort and make unique.
+    auto axesValue = node->namedInput(attr::dim);
+    for (auto axisNode : axesValue->node()->inputs()) {
```

Comment:
I think we're expecting to see `prim::ListConstruct` here and we're iterating over its inputs, which are expected to be `prim::Constant`s? Could we add some asserts to actually verify that?

Reply:
I've added an assert to check it's `prim::ListConstruct`. Subsequent casts of elements to `IntImm` should take care of asserting the rest.